File: TextosJapones/src/main/java/org/jisho/textosJapones/controller/mangas/MangasVincularCelulaPequenaController.java
package org.jisho.textosJapones.controller.mangas;

import javafx.event.EventHandler;
import javafx.fxml.FXML;
import javafx.fxml.Initializable;
import javafx.scene.image.ImageView;
import javafx.scene.input.ClipboardContent;
import javafx.scene.input.Dragboard;
import javafx.scene.input.MouseEvent;
import javafx.scene.input.TransferMode;
import javafx.scene.layout.AnchorPane;
import javafx.scene.layout.VBox;
import javafx.scene.text.Text;
import org.jisho.textosJapones.components.listener.VinculoListener;
import org.jisho.textosJapones.model.entities.VinculoPagina;
import org.jisho.textosJapones.model.enums.Pagina;
import org.jisho.textosJapones.util.Util;

import java.net.URL;
import java.util.ResourceBundle;

public class MangasVincularCelulaPequenaController implements Initializable {

	@FXML
	public AnchorPane root;

	@FXML
	public ImageView imagem;

	@FXML
	public Text numero;

	@FXML
	public VBox containerNome;

	@FXML
	public Text nomePagina;

	private VinculoListener listener;

	public void setDados(VinculoPagina vinculo) {
		imagem.setImage(vinculo.getImagemVinculadoEsquerda());
		numero.setText(vinculo.getVinculadoEsquerdaPagina().toString());
		nomePagina.setText(vinculo.getVinculadoEsquerdaNomePagina());

		root.setOnDragDetected(new EventHandler<MouseEvent>() {
			public void handle(MouseEvent event) {
				listener.onDragStart();
				Dragboard db = root.startDragAndDrop(TransferMode.ANY);

				db.setDragView(Util.criaSnapshot(root), event.getX(), event.getY());

				vinculo.onDragOrigem = Pagina.NAO_VINCULADO;
				ClipboardContent content = new ClipboardContent();
				content.put(Util.VINCULO_ITEM_FORMAT, vinculo);
				content.put(Util.NUMERO_PAGINA_ITEM_FORMAT, vinculo.getVinculadoEsquerdaPagina().toString());
				content.putString(vinculo.getVinculadoEsquerdaNomePagina());
				db.setContent(content);

				event.consume();
			}
		});
	}

	public void setListener(VinculoListener listener) {
		this.listener = listener;
	}

	@Override
	public void initialize(URL location, ResourceBundle resources) {

		imagem.fitWidthProperty().bind(root.widthProperty());
		imagem.fitHeightProperty().bind(root.heightProperty());
		imagem.setPreserveRatio(true);
		imagem.fitWidthProperty().bind(containerNome.widthProperty());

	}

	public static URL getFxmlLocate() {
		return MangasVincularCelulaPequenaController.class.getResource("/view/mangas/MangaVincularCelulaPequena.fxml");
	}
}

File: ProcessaTexto/src/main/kotlin/br/com/fenix/processatexto/controller/mangas/MangasVincularCelulaPequenaController.kt
package br.com.fenix.processatexto.controller.mangas

import br.com.fenix.processatexto.components.listener.VinculoListener
import br.com.fenix.processatexto.model.entities.processatexto.VinculoPagina
import br.com.fenix.processatexto.model.enums.Pagina
import br.com.fenix.processatexto.util.Utils
import javafx.fxml.FXML
import javafx.fxml.Initializable
import javafx.scene.image.ImageView
import javafx.scene.input.ClipboardContent
import javafx.scene.input.TransferMode
import javafx.scene.layout.AnchorPane
import javafx.scene.layout.VBox
import javafx.scene.text.Text
import java.net.URL
import java.util.*


class MangasVincularCelulaPequenaController : Initializable {

    @FXML
    lateinit var root: AnchorPane

    @FXML
    lateinit var imagem: ImageView

    @FXML
    lateinit var numero: Text

    @FXML
    lateinit var containerNome: VBox

    @FXML
    lateinit var nomePagina: Text

    private lateinit var listener: VinculoListener
    fun setDados(vinculo: VinculoPagina) {
        imagem.image = vinculo.imagemVinculadoEsquerda
        numero.text = vinculo.vinculadoEsquerdaPagina.toString()
        nomePagina.text = vinculo.vinculadoEsquerdaNomePagina
        root.setOnDragDetected {
            listener.onDragStart()
            val db = root.startDragAndDrop(TransferMode.COPY, TransferMode.MOVE, TransferMode.LINK)
            db.setDragView(Utils.criaSnapshot(root), it.x, it.y)
            vinculo.onDragOrigem = Pagina.NAO_VINCULADO
            val content = ClipboardContent()
            content[Utils.VINCULO_ITEM_FORMAT] = vinculo
            content[Utils.NUMERO_PAGINA_ITEM_FORMAT] = vinculo.vinculadoEsquerdaPagina.toString()
            content.putString(vinculo.vinculadoEsquerdaNomePagina)
            db.setContent(content)
            it.consume()
        }
    }

    fun setListener(listener: VinculoListener) {
        this.listener = listener
    }

    @Override
    override fun initialize(location: URL?, resources: ResourceBundle?) {
        imagem.fitWidthProperty().bind(root.widthProperty())
        imagem.fitHeightProperty().bind(root.heightProperty())
        imagem.isPreserveRatio = true
        imagem.fitWidthProperty().bind(containerNome.widthProperty())
    }

    companion object {
        val fxmlLocate: URL get() = MangasVincularCelulaPequenaController::class.java.getResource("/view/mangas/MangaVincularCelulaPequena.fxml") as URL
    }
}